pkg/kubelet/container/interface.go

```diff
@@ -0,0 +1,365 @@
+	// Create creates a container in the sandbox, and returns the ID
+	// of the created container.
+	Create(config *ContainerConfig, sandboxConfig *PodSandboxConfig, sandboxID PodSandboxID) (RawContainerID, error)
+	// Start starts a created container.
+	Start(id RawContainerID) error
+	// Stop stops a running container with a grace period (i.e., timeout).
+	Stop(id RawContainerID, timeout int) error
+	// Remove removes the container.
+	Remove(id RawContainerID) error
+	// List lists the existing containers that match the ContainerFilter.
+	// The returned list should only include containers previously created
+	// by this ContainerManager.
+	List(filter ContainerFilter) ([]Container, error)
+	// Status returns the status of the container.
+	Status(id RawContainerID) (RawContainerStatus, error)
+	// Exec executes a command in the container.
+	Exec(id RawContainerID, cmd []string, streamOpts StreamOptions) error
+}
+
+type ContainerConfig struct {
+	// Name of the container. The string should conform to [a-zA-Z0-9_-]+.
+	Name string
+	// Image to use.
+	Image ImageSpec
+	// Command to execute (i.e., entrypoint for docker)
+	Command []string
+	// Args for the Command (i.e., command for docker)
+	Args []string
+	// Current working directory of the command.
+	WorkingDir string
+	// List of environment variable to set in the container
+	Env []KeyValue
+	// Mounts specifies mounts for the container
+	Mounts []Mount
+	// Labels are key value pairs that may be used to scope and select individual resources.
+	Labels Labels
+	// Annotations is an unstructured key value map that may be set by external
+	// tools to store and retrieve arbitrary metadata.
+	Annotations map[string]string
+	// Privileged runs the container in the privileged mode.
+	Privileged bool
+	// ReadOnlyRootFS sets the root filesystem of the container to be
+	// read-only.
+	ReadOnlyRootFS bool
+	// Path relative to PodSandboxConfig.LogDirectory for container to store
+	// the log on the host.
+	LogPath string
+
+	// Variables for interactive containers, these have very specialized
+	// use-cases (e.g. debugging).
+	// TODO: Determine if we need to continue supporting these fields that are
+	// part of Kubernetes's Container Spec.
+	STDIN     bool
+	STDINONCE bool
+	TTY       bool
+
+	// Linux contains configuration specific to Linux containers.
+	Linux *LinuxContainerConfig
+}
+
+// RawContainerStatus represents the status of a container.
+type RawContainerStatus struct {
+	// ID of the container.
+	ID ContainerID
+	// Name of the container.
+	Name string
+	// Status of the container.
+	State ContainerState
+	// Creation time of the container.
+	CreatedAt unversioned.Time
+	// Start time of the container.
+	StartedAt unversioned.Time
+	// Finish time of the container.
+	FinishedAt unversioned.Time
+	// Exit code of the container.
+	ExitCode int
+	// Name of the image, this also includes the tag of the image,
+	// the expected form is "NAME:TAG".
+	Image string
```

Comment:
The `Image` and `ImageID` fields, AFAIK, are only used in the `PodStatus` for inspection purpose. From a user's perspective, returning everything in [`Image`](https://github.com/kubernetes/kubernetes/blob/423a4154beab22649735e32283a045b5cd574832/pkg/kubelet/container/runtime.go#L317) seems excessive. I think it's reasonable to return an ID (or digest) to uniquely identify an image, together with a string that the user specifies as the image the container spec. This can be achieved by writing the string as an annotation in the container.

As for what the runtime should return (as opposed to what the k8s api should expose to the end user), ideally, I think it should return simply the image ID or just the rootfs path. Kubelet should be the one piecing the information together. Before kubelet can do that, it will need the container runtime to provide the `ImageID` associated with the container. We could reuse `Image` (the struct), but it provides too much arbitrarily information (`RepoTags` and `RepoDigests`) that we don't need for the container. What if we just return `ImageID` and get the `Image` string from the annotation? 

Also /cc @vishh who has stronger opinions on images.

By the way, the "name:tag" description is actually not correct. In docker's case, the `Image` string will be whatever the user specified, e.g., "busybox"